File: client/src/boundary/interactor/GetGameInteractor.ts
import { GetGameUC } from "boundary/api/GetGameUC";
import GameStorage from "services/api/GameStorage";
import BoundaryGame from "boundary/model/BoundaryGame";
import { GameD2BConverter } from "./GameD2BConverter";
import { map, Observable } from "rxjs";

export class GetGameInteractor implements GetGameUC {
  private gameStorage: GameStorage;
  private gameConverter: GameD2BConverter;

  constructor(gameStorage: GameStorage, gameConverter: GameD2BConverter) {
    this.gameStorage = gameStorage;
    this.gameConverter = gameConverter;
  }

  getGame(id: string): Observable<BoundaryGame> {
    return this.gameStorage
      .getGame(id)
      .pipe(map((game) => this.gameConverter.convert(game)));
  }
}

File: server/battleships/use_cases/implementation/src/main/java/eu/optas/use_cases/implementation/GetGameInteractor.java
package eu.optas.use_cases.implementation;

import eu.optas.domain.Game;
import eu.optas.gateway.api.GameGateway;
import eu.optas.use_cases.api.BoundaryGame;
import eu.optas.use_cases.api.GetGameUC;

import java.util.Optional;

public class GetGameInteractor implements GetGameUC {
    private final GameGateway gameGateway;
    private final GameD2BConverter gameD2BConverter;

    public GetGameInteractor(GameGateway gameGateway, GameD2BConverter gameD2BConverter) {
        this.gameGateway = gameGateway;
        this.gameD2BConverter = gameD2BConverter;
    }

    @Override
    public Optional<BoundaryGame> getGame(String id) {
        Game game = gameGateway.getGame(id);
        return Optional.ofNullable(game).map(gameD2BConverter::convert);
    }
}
